### PR TITLE
Synchronized collection rebase

### LINKED ISSUE #404: Update from Version: 0.7.1 -> 0.8.0 fails
We are trying to update to the new Version `0.8.0` via the [update script](https://github.com/ComPlat/chemotion_ELN/blob/development/scripts/update_production.sh). I haven't altered the script and run it as is.

It seems to have an issue with the database migrations:

```
== 20201109012155 CreateTextTemplates: migrating ==============================

-- create_table(:text_templates)

001:29 removing tmp files..
001:29 removing tmp repo..
001:29 An error has occured
```

Looking more closley
```
00:25 deploy:migrating
01 /tmp/chemotion/nvm-exec.sh ~/.rvm/bin/rvm 2.6.6 do bundle exec rake db:migrate
01 Now using node v12.22.1 (npm v7.11.1)
01 /home/production/.rvm/gems/ruby-2.6.6/gems/activesupport-4.2.11.3/lib/active_support/core_ext/object/duplicable.rb:111: warning: BigDecimal.new is deprecated; use BigDecimal() method instead.
01 rake aborted!
01 StandardError: An error has occurred, this and all later migrations canceled:
01 
01 PG::DuplicateTable: ERROR:  relation "text_templates" already exists
01 : CREATE TABLE "text_templates" ("id" serial primary key, "type" character varying, "user_id" integer NOT NULL, "name" character varying, "data" jsonb DEFAULT '{}', "deleted_at" timestamp, "created_at" timestamp NOT NULL, "updated_at" timestamp NOT NULL) 

```
[...]
```
== 20201109012155 CreateTextTemplates: migrating ==============================
-- create_table(:text_templates)
rake stderr: /home/production/.rvm/gems/ruby-2.6.6/gems/activesupport-4.2.11.3/lib/active_support/core_ext/object/duplicable.rb:111: warning: BigDecimal.new is deprecated; use BigDecimal() method instead.
rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

PG::DuplicateTable: ERROR:  relation "text_templates" already exists
: CREATE TABLE "text_templates" ("id" serial primary key, "type" character varying, "user_id" integer NOT NULL, "name" character varying, "data" jsonb DEFAULT '{}', "deleted_at" timestamp, "created_at" timestamp NOT NULL, "updated_at" timestamp NOT NULL) 

```
it looks like, it doesn't try to migrate but recreate the entire database.
I attached the full log.
Do I have to make some changes to the update script?

Best
Fabian

[update_production.log](https://github.com/ComPlat/chemotion_ELN/files/6447636/update_production.log)

### LINKED ISSUE #438: ELN adminstration/ generic elements: Create field types for general drag and drop function
At the moment, it is only possible to have field types of drag and drop function for Samples and Molecules, namely: "Drag Sample" and "Drag Molecule". It would be great to have those field types with general scope, so to have only a "drag and drop" field type that can be applied on any element.

### PR DESCRIPTION
**Pull request by PiTrem**
Synchronised an owned collection with users or groups but unlike sharing, the users see the same original collection.
permissions are stored in the associating sync_collections_users model
UI management thru collection management

resolve #404
resolve #438 #414 #428 #431 #432

### CODE REVIEW

File: app/api/chemotion/sync_collection_api.rb [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+#TODO
+      namespace :take_ownership do
+        desc "Take ownership of collection with specified sync_collections_user id"
+         params do
+           requires :id, type: Integer, desc: "SyncCollectionsUSer id"
+         end
+         route_param :id do
+           before do
+      #       error!('401 Unauthorized', 401) unless CollectionPolicy.new(current_user, Collection.find(params[:id])).take_ownership?
```
Comment (Sergeykot) [inline app/api/chemotion/sync_collection_api.rb]: commented code

File: app/api/chemotion/sync_collection_api.rb [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+           end
+
+           post do
+      #       Usecases::Sharing::TakeOwnership.new(params.merge(current_user_id: current_user.id)).execute!
```
Comment (Sergeykot) [inline app/api/chemotion/sync_collection_api.rb]: commented code

File: app/api/chemotion/sync_collection_api.rb [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        before do
+          c = Collection.where(is_shared:false,id: params[:id],user_id: current_user.id).first
+          if c
+            samples =   c.samples
+            reactions = c.reactions
+            wellplates = c.wellplates
+            screens = c.screens
+
+            top_secret_sample = samples.pluck(:is_top_secret).any?
+            top_secret_reaction = reactions.flat_map(&:samples).map(&:is_top_secret).any?
```
Comment (Sergeykot) [inline app/api/chemotion/sync_collection_api.rb]: could it be maybe optimized to ActiveRecord query with '.includes' and '.pluck' methods

File: app/assets/javascripts/components/CollectionSubtree.js [hunk 4]
```diff
@@ -179,4 +193,8 @@ CollectionSubtree.propTypes = {
   selected: React.PropTypes.bool,
   root: React.PropTypes.object,
   visible: React.PropTypes.bool,
+  // isSync: React.PropTypes.bool,
```
Comment (Sergeykot) [inline app/assets/javascripts/components/CollectionSubtree.js]: commented code

File: app/assets/javascripts/components/stores/UIStore.js [hunk 5]
```diff
@@ -178,6 +181,33 @@ class UIStore {
     }
   }
 
+  handleSelectSyncCollection(collection) {
+    // TODO
```
Comment (Sergeykot) [inline app/assets/javascripts/components/stores/UIStore.js]: not clear to which place this TODO applies

File: spec/api/sync_collection_api_spec.rb [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+  context 'when authorized user logged in' do
+
+    before do
+      allow_any_instance_of(WardenAuthentication).to receive(:current_user).and_return(u1)
+    end
+
+    describe 'GET /api/v1/syncCollections/:id, ' do
+      context 'on a outgoing sync_collection,' do
+        before do
+      #    sc1_2.save!
```
Comment (Sergeykot) [inline spec/api/sync_collection_api_spec.rb]: is that line needed?

File: spec/api/sync_collection_api_spec.rb [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+        end
+        it 'does not return the sync_collections_user' do
+          response_body = JSON.parse response.body
+          expect(response.status).to be 404
+          expect(response_body['error']).to_not be_empty
+        end
+      end
+      context ', on a incoming sync_collection,' do
+        before do
+      #    sc2_1.save!
```
Comment (Sergeykot) [inline spec/api/sync_collection_api_spec.rb]: is this lien also needed?

**Comment by Sergeykot** [comment]
comments left

File: app/api/chemotion/sync_collection_api.rb (hunk 3 above)
Comment (PiTrem) [inline app/api/chemotion/sync_collection_api.rb]: same issue a for collection_api (from which the part  is copied)
I ll make a new item : ELN collection logic => ensure a collection always contains the samples used in its reactions/wellplates/screens ( collection_api_spec)
Then we ll only have to check on the collection samples

**Pull request merged**